src/main/java/edu/immutables/bst/Branch.java

```diff
@@ -0,0 +1,98 @@
+package edu.immutables.bst;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Representation of Inner node in the tree.
+ *
+ * @param <T> type of elements held by the tree
+ * @author tarek-nawara
+ */
+class Branch<T extends Comparable<T>> implements Tree<T> {
+    private final Tree<T> left;
+    private final Tree<T> right;
+    private final T value;
+    private final int size;
+
+    Branch(final T value, final Tree<T> left, final Tree<T> right) {
+        this.value = value;
+        this.left = left;
+        this.right = right;
+        this.size = 1 + left.size() + right.size();
+    }
+
+    @Override
+    public Tree<T> add(T element) {
+        int cmp = element.compareTo(value);
+        if (cmp < 0) return new Branch<>(value, left.add(element), right);
+        else if (cmp > 0) return new Branch<>(value, left, right.add(element));
+        return this;
```

Comment:
Why are you making the elements unique in the bst ?
The BST is mainly used as a multiset, is it for the ImmutableMap ?

Reply:
One of the models for Binary search trees is that it doesn't contain any duplicate elements, you can see these model here: 
https://www.hackerrank.com/challenges/is-binary-search-tree/problem